### ISSUE TITLE
Consider raising when a jitted function's output is not PyTree-like

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#2854]
Just as https://github.com/google/jax/issues/2813 argues that inputs to `jit` should always be hashable since that prevents unnecessary recompilation, another common error with `jit` is when its output is not PyTree-like (either a PyTree or registered using `register_pytree_node`).  This has caught me a few times with tracers leaking out of functions.  It makes me apprehensive about decorating functions with `jit`.  Is there any use case for non-PyTree-like objects being returned?  If not, I propose checking that the return value is a PyTree and raising if not.

**Comment by gnecula** [comment]
Tracers leaking out of functions seems to be a separate issue from return values not being PyTrees. Are these leaking tracers returned through global state? Can you show an example of the error that you encountered?

**Comment by NeilGirdhar** [comment]
Oh you mean like a real code example?  Here's the last one that hit me.  I have a class that holds parameters, and which uses a while loop to generate a trajectory:

```
    def infer_trajectory(self,
                         eif: EncodingIteratedFunction) -> Tuple[
                             EncodingConfiguration, Trajectory]:
        """
        Returns:
            final_rec: The final EncodingConfiguration.
            trajectory: The trajectory.
        """
        state, recs = eif.sample_trajectory(
            self, self._initial_configuration())
        trajectory = Trajectory(recs)
        return state.current_x, trajectory
```
If this function is jitted, then the values that are copied out of recs and into `Trajectory` are tracers.  If I register `Trajectory` as a PyTree node, then these tracers will automatically be decoded.

My question in this is issue is why would I ever want to return something that's not a PyTree node from a jitted function?  In this example, it was just a source of bugs.

**Comment by jekbradbury** [comment]
This is interesting. If I'm following correctly, we already do the same check for function _arguments_ (we throw a "not a JaxType" error on unknown PyTree leaves). It seems like it would be sensible to also error on unknown PyTree leaves in return values, but there also might be valid uses for this (basically returning "compile-time" Python data).

**Comment by NeilGirdhar** [comment]
@jekbradbury Right, that makes sense.

**Comment by NeilGirdhar** [comment]
@jekbradbury Right, good point, that makes sense.

Maybe when the caller wants to return compile-time data, they should be explicit that that's what they're doing?  After all, I think it's a pretty rare thing to do.  I don't know what sort of compile-time data is available, but it could all be wrapped into an object like: `jax.CompileTimeData` or something?

An object like this has the advantage that it could contain all kinds of internal requests and assertions that wouldn't have to be resolved until the jitting mechanism at the end of the function gets ahold of it.  The result of the compile-time data could be returned to the first caller of the function, or every caller, or none of the callers and instead be set as an attribute on the function itself?

I don't know, but I feel like this rare case of returning compile-time data should be explicitly requested?

**Comment by gnecula** [comment]
I think I understand. The following code 
```
class MyClass:
  def __init__(self, x):
     self.x = x

jax.jit(lambda x: MyClass(x + 2))(0)
```

will return an instance of `MyClass` with the tracer inside. No code is jitted, because JAX does not see the returned Tracers. This is a confusing.

I think that James is thinking about an example like:
```
jax.jit(lambda x: x + 5, "hello")(0)
```

Maybe for such examples we can use a `has_aux` parameter to the jit to declare the the function being jitted as an intentional auxiliary result (a la [grad](https://jax.readthedocs.io/en/latest/jax.html#jax.grad))

**Comment by NeilGirdhar** [comment]
This looks like it's been done!
